photon-ml/src/main/scala/com/linkedin/photon/ml/data/AvroDataReader.scala

```diff
@@ -0,0 +1,375 @@
+
+    // Retrieve the features
+    val features = readFeaturesFromRecord(record, fieldNames)
+
+    // Check for duplicate features
+    val duplicateFeatures = features
+      .groupBy(_._1)
+      .filter(_._2.length > 1)
+      .map { case (k, v) => (k, v.map(_._2).toList) }
+    require(duplicateFeatures.isEmpty, s"Duplicate features found: ${duplicateFeatures.toString}")
+
+    // Resolve feature keys to indices
+    val featuresWithIndices = features.flatMap { case (featureKey, value) =>
+      if (featureMap.contains(featureKey)) {
+        Some(featureMap.getIndex(featureKey) -> value)
+      } else {
+        None
+      }
+    }
+
+    // Add intercept if necessary
+    val addIntercept = featureMap.contains(INTERCEPT_KEY)
+    val featuresWithIntercept = if (addIntercept) {
+      featuresWithIndices ++ Array(featureMap.getIndex(INTERCEPT_KEY) -> 1.0)
+    } else {
+      featuresWithIndices
+    }
+
+    // Create feature vector
+    VectorUtils.breezeToMllib(
+      VectorUtils.convertIndexAndValuePairArrayToSparseVector(
+        featuresWithIntercept, featureMap.featureDimension))
+  }
+
+  /**
+   * Infers spark sql field schema by sampling a single record from the RDD of avro GenericRecord.
+   *
+   * @param records the avro records
+   * @return the spark sql field schema
+   */
+  def inferSchemaFields(records: RDD[GenericRecord]): Seq[StructField] = records
+    .map(r => r.getSchema
+      .getFields
+      .asScala
+      .flatMap { f => avroTypeToSql(f.name, f.schema) })
+    .take(1)
+    .head
+
+  /**
+   * Converts the named avro field schema to an equivalent spark sql schema type.
+   *
+   * @param name the field name
+   * @param avroSchema the avro schema for the field
+   * @return spark sql schema for the field
+   */
+  def avroTypeToSql(name: String, avroSchema: Schema): Option[StructField] = avroSchema.getType match {
+    case INT => Some(StructField(name, IntegerType, nullable = false))
+    case STRING => Some(StructField(name, StringType, nullable = false))
+    case BOOLEAN => Some(StructField(name, BooleanType, nullable = false))
+    case DOUBLE => Some(StructField(name, DoubleType, nullable = false))
+    case FLOAT => Some(StructField(name, FloatType, nullable = false))
+    case LONG => Some(StructField(name, LongType, nullable = false))
+    case MAP =>
+      avroTypeToSql(name, avroSchema.getValueType).map { valueSchema =>
+        StructField(
+          name,
+          MapType(StringType, valueSchema.dataType, valueContainsNull = valueSchema.nullable),
+          nullable = false)
+      }
+
+    case UNION =>
+      if (avroSchema.getTypes.asScala.exists(_.getType == NULL)) {
+        // In case of a union with null, take the first non-null type for the value type
+        val remainingUnionTypes = avroSchema.getTypes.asScala.filterNot(_.getType == NULL)
+        avroTypeToSql(name, remainingUnionTypes.head).map(_.copy(nullable = true))
```

Comment:
Yes, this only handles union fields for the case where the type has been unioned with "null" (we see this pretty often for optional fields). Any other kind of union is ignored. We could do something a little more sophisticated like you describe, where we maintain a table of "unionable" types and read it into a predetermined gcd type.